### ISSUE TITLE
GPU-writes to persistent mapped buffers have no effect

### ISSUE THREAD

**Issue by schmorf** [baldurk/renderdoc#677]
I am using a persistent mapped buffer as a counter. I reset the counter to some arbitrary number, call a compute shader where i modify the buffer and after that the counter still has the same value when I run my application in RenderDoc. Outside of RenderDoc it works as expected. It seems like buffer writes in shaders are simply ignored when I run it in RenderDoc.
In a minimal example I create and map the buffer like so:
```
glCreateBuffers(1, &atomicCounterBuffer);
GLbitfield flags = GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT;
glNamedBufferStorage(atomicCounterBuffer, 4, nullptr, flags);
glBindBufferBase(GL_ATOMIC_COUNTER_BUFFER, 0, atomicCounterBuffer);
pCounter = static_cast<uint32_t*>(glMapNamedBufferRange(atomicCounterBuffer, 0, 4, flags));
```
and then call my compute shader:
```
*pCounter = 0;
glDispatchCompute(1, 1, 1);
glFinish();
std::cout << *pCounter << '\n';
```
The shader just does a single increment:
```
#version 450 core
layout(binding = 0, offset = 0) uniform atomic_uint atomicCounter;
layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;
void main()
{
   atomicCounterIncrement(atomicCounter);
}
```
It should print '1' but when I launch it in RenderDoc I get '0'. The same happens with shader storage buffer objects.
Oddly enough when I actually do a capture and view the buffer contents I can see the correct value.
Windows 10
GTX 1070 with 382.53

**Comment by baldurk** [comment]
Ah, my persistent mapped buffer handling only handles CPU-writes/GPU-reads, not GPU-writes/CPU-reads.

I'll need to add a check to propagate data in the other direction, which might not be cheap. Although I've also been thinking about changing the way I handle persistent mapped buffers to the way I do it in vulkan which is much more efficient - I'm just not sure if I can do it legally in GL.

**Comment by devshgraphicsprogramming** [comment]
Question, if I were to suggest an option to disable persistent coherent buffer change tracking, would that be its own issue?

Something like a "slim capture mode" where buffers would display warnings that the captured data may not be correct?

**Comment by baldurk** [comment]
Disabling coherent buffer interception would almost certainly cause broken captures, so I don't think it would be a good option to have. It would be very easy for users to enable the option and cause corruption without understanding why.

In the niche case where an application is persistently mapping buffers that it doesn't actually change it would be better to add a check in the application itself to detect renderdoc and switch to regular map/unmap. Indeed persistent maps aren't free even without renderdoc, so if the data isn't changing it would likely be more optimal.

**Comment by devshgraphicsprogramming** [comment]
You could have it as an advanced capture option, and not consider bug reports on captures that have it turned on.

On the persistent buffers themselves:
1) They seem to be free or have <1% overhead in my benchmarks on Intel HD Graphics.
(I dont have to do staging buffers when streaming buffer data, at all)
Map/Unmap would actually cause 2 data copies, additional driver thread work and a sync here.

2) Moreover they are reported to be the fastest when using (Multi)DrawIndirect and not generating data on the GPU.

3) One could switch to regular map/unmap but sometimes thats not an option for the general architecture of the underlying framework/engine, like when one uses circular buffers for a very dynamic, fragmented and 3rdparty GUI. 

P.S. I actually only use Persistent Buffers for GUI, instanced node staging buffers, some UBOs and L-Buffer sub-image download on Nvidia/AMD .

**Comment by baldurk** [comment]
It would still be very easy for someone to enable the option without a full understanding and end up with things not working in a very non-obvious way. I'm not too concerned about bug reports because then it's possible to diagnose and let them know the problem, I'm more concerned with someone who finds things are broken and gives up, thinking RenderDoc is broken.

What you're describing sounds like you are uploading data through the map, which means that you would also get broken captures with this option, so I'm not sure I really understand what you're proposing. Completely disabling persistent map tracking means threading the needle through a very precise set of circumstances where it wouldn't break anything.

Another application-side change you can make if you are uploading data through maps but don't want the heavy overhead is to use `GL_MAP_FLUSH_EXPLICIT_BIT` instead of `GL_MAP_COHERENT_BIT` since then no heavyweight tracking is needed and renderdoc can just record the flushes.

**Comment by devshgraphicsprogramming** [comment]
> Another application-side change you can make if you are uploading data through maps but don't want the heavy overhead is to use GL_MAP_FLUSH_EXPLICIT_BIT instead of GL_MAP_COHERENT_BIT since then no heavyweight tracking is needed and renderdoc can just record the flushes.

Thats awesome, I'll go and do that then.

**Comment by devshgraphicsprogramming** [comment]
We use flush explicit and map coherent, but we have a  capture+runtime bug.

Currently it breaks the programs that are running through RenderDoc, before even a capture gets done.
I.e. without renderdoc a mesh rendered from persistently mapped memory shows up, without it doesn't.

Upon examination in render doc the buffer's values are always 0xdddddddd despite writing and flushing.

**Comment by baldurk** [comment]
Please open a new issue for your problem with a repro case if possible. This issue is specifically to track the lack of GPU-writes to persistent mapped buffers in GL, it's not a general issue on all things related to persistent mapped buffers in GL.

**Comment by devshgraphicsprogramming** [comment]
Will get the person experiencing the issues to do so, we actually have a minimal reproducible example too.

**Comment by manhnt9** [comment]
Issue posted.

**Comment by baldurk** [comment]
I'm going to close this down as I think GL persistent map handling should be much better now and the original issue should be solved. I've refactored it to be much closer to the vulkan map imlementation - the map isn't intercepted and instead during capture at any sync point the contents are read back. This means that the `READ` flag is added onto any persistent map/buffer, which shouldn't change behaviour for correctly functioning programs but may have a slight perf impact on most GL drivers.

The maps should then be be significantly faster while idle - instead of potentially slowing down to many 100s of milliseconds per frame there should be only whatever impact comes from changing a write-only map to a read-write map.

During capture, any time a sync point is hit the current contents of the map are diffed against the last known contents, and any change is saved out just before that call. This does mean a capture can be quite slow, as the time taken scales with `O(M*N)` where `M` is the number of bytes of mapped memory, and `N` is the number of sync points (i.e. draw calls).

Note that non-coherent flush-explicit maps will still be much more efficient as on GL the number of sync points during capture is still very high (at least every draw, and a number of other functions that could potentially read from a buffer).

tl;dr persistent maps will always return the original pointer from GL to the user, regardless of flags, which should solve the original issue as there's no shadowing going on with renderdoc that needs to be updated.
